subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/moduleconverter/dependencies/DefaultDependencyDescriptorFactoryTest.groovy

```diff
@@ -56,10 +58,34 @@ public class DefaultDependencyDescriptorFactoryTest extends Specification {
         and:
         def dependencyDescriptorFactory = new DefaultDependencyDescriptorFactory(
                 ivyDependencyDescriptorFactory1
-        );
+        )
         dependencyDescriptorFactory.createDependencyDescriptor(componentId, configurationName, null, projectDependency)
 
         then:
         thrown InvalidUserDataException
     }
+
+    def "creates descriptor for dependency constraints"() {
+        given:
+        def dependencyConstraint = new DefaultDependencyConstraint("g", "m", "1")
+
+        when:
+        def dependencyDescriptorFactory = new DefaultDependencyDescriptorFactory()
+        def created = dependencyDescriptorFactory.createDependencyConstraintDescriptor(componentId, configurationName, null, dependencyConstraint)
+        def selector = created.selector as ModuleComponentSelector
+
+        then:
+        created.optional
```

Comment:
This is the kind of things that will become confusing for future readers. Why would a constraint be optional?

Reply:
Will also be fixed with #3626